lib/Analysis/Utility.cpp

```diff
@@ -329,6 +331,230 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+
+  // Now we have C' = P2^-1 ∘ C = W ∘ P1. W is considerably easier to compute.
+  // A warp shuffle is a function from `(register, lane) -> (lane)`, i.e.
+  //
+  //   W = [ I R' ]
+  //       [ 0 L  ]
+  //
+  // `W^-1 ∘ C'` will be
+  //
+  //   [ I R ] [ C'(r,r) C'(r,l) ] = [ ... C'(r,l) + R*C'(l,l) ]
+  //   [ 0 L ] [ C'(l,r) C'(l,l) ] = [ ... L*C'(l,l)           ]
+  //
+  // Since P1 cannot change lanes, we know that
+  //
+  //   W^-1 ∘ C' = [ ... 0 ]
+  //               [ ... I ]
+  //
+  // Thus L = C'(l,l)^-1, and R = -C'(r,l) * C'(l,l)^-1. (0 - LL) = LL in GF(2).
+  // We know that C'(l,l) has a suitable pseudo-inverse.
+  LinearLayout L = Cp.sublayout(kLane, kLane).pseudoinvert();
+  LinearLayout R = Cp.sublayout(kRegister, kLane).compose(L);
+
+  // Now form W^-1.
+  LinearLayout WinvLeft =
+      LinearLayout::identity1D(dstRegSize, kRegister, kRegister)
+          .concatIns(
+              LinearLayout::zeros1D(numLanes, kLane, kRegister, dstRegSize));
+  LinearLayout Winv = WinvLeft.concatOuts(R.concatIns(L));
+
+  // Check that Winv was formed correctly. P1 is just what's left over.
+  LinearLayout P1 = Winv.compose(Cp);
+  assert(P1.sublayoutIsZero(kRegister, kLane));
+  assert(squareSublayoutIsIdentity(P1, kLane));
+
+  // Grab just the interesting parts of the decomposed layouts.
+  P1 = P1.sublayout({kLane, kRegister}, kRegister);
+  P2inv = P2inv.sublayout({kLane, kRegister}, kRegister);
+  Cp = Cp.sublayout({kLane, kRegister}, kLane);
+
+  // To minimize the number of selects emitted on the source side, determine the
+  // minimum set of registers that could be selected from each thread.
+  // InstCombine *might* be able to crush this, but if the sizePerThread is
+  // large, it's truly a huge number of selects that get emitted.
+  // If reducedP1 is trivial, then we will emit
+  // shflSrc = select(i == i, src[i], undef) and this will get trivially folded,
+  // so don't worry about this case.
+  LinearLayout reducedP1 = P1.removeZeroBasesAlongDim(kLane);
+  LinearLayout reducedP2 = P2inv.removeZeroBasesAlongDim(kLane);
+
+  // The number of emitted selects can still be quite large if the layout is not
+  // cooperative. This happens when the source register is more correlated
+  // with the desination lane than the destination register (i.e. the number of
+  // non-zero bases). The number of selects impacts performance and grows
+  // exponentially with the number of non-zero bases. Experiments show that more
+  // than 1 select causes performance to be slower than shared memory.
+  if (reducedP1.getInDimSize(kLane) > 2 || reducedP2.getInDimSize(kLane) > 2)
+    return {};
+
+  // HACK: Workaround AMD codegen path generating transient invalid layouts.
+  auto isInvalidDotEnc = [](RankedTensorType type) {
+    auto dotEnc = dyn_cast<DotOperandEncodingAttr>(type.getEncoding());
+    return dotEnc && dotEnc.getKWidth() == 0;
+  };
+  if (isInvalidDotEnc(srcTy) || isInvalidDotEnc(dstTy))
+    return {};
```

Comment:
This is fine, but you can be more explicit by seeing if the parent is not an NvidiaMmaLayout (or a Blocked layout... does the FMA layout work?).
cc @antiagainst for the AMD skips

Reply:
Added a check for AMDMfma/Wmma attr